numba/cuda/mathimpl.py

```diff
@@ -0,0 +1,246 @@
+import math
+import operator
+from llvmlite import ir
+from numba.core import types, typing
+from numba.core.imputils import Registry
+from numba.types import float32, float64, int64, uint64
+from numba.cuda import libdevice
+
+registry = Registry()
+lower = registry.lower
+
+
+booleans = []
+booleans += [('isnand', 'isnanf', math.isnan)]
+booleans += [('isinfd', 'isinff', math.isinf)]
+booleans += [('isfinited', 'finitef', math.isfinite)]
+
+unarys = []
+unarys += [('ceil', 'ceilf', math.ceil)]
+unarys += [('floor', 'floorf', math.floor)]
+unarys += [('fabs', 'fabsf', math.fabs)]
+unarys += [('exp', 'expf', math.exp)]
+unarys += [('expm1', 'expm1f', math.expm1)]
+unarys += [('erf', 'erff', math.erf)]
+unarys += [('erfc', 'erfcf', math.erfc)]
+unarys += [('tgamma', 'tgammaf', math.gamma)]
+unarys += [('lgamma', 'lgammaf', math.lgamma)]
+unarys += [('sqrt', 'sqrtf', math.sqrt)]
+unarys += [('log', 'logf', math.log)]
+unarys += [('log10', 'log10f', math.log10)]
+unarys += [('log1p', 'log1pf', math.log1p)]
+unarys += [('acosh', 'acoshf', math.acosh)]
+unarys += [('acos', 'acosf', math.acos)]
+unarys += [('cos', 'cosf', math.cos)]
+unarys += [('cosh', 'coshf', math.cosh)]
+unarys += [('asinh', 'asinhf', math.asinh)]
+unarys += [('asin', 'asinf', math.asin)]
+unarys += [('sin', 'sinf', math.sin)]
+unarys += [('sinh', 'sinhf', math.sinh)]
+unarys += [('atan', 'atanf', math.atan)]
+unarys += [('atanh', 'atanhf', math.atanh)]
+unarys += [('tan', 'tanf', math.tan)]
+unarys += [('tanh', 'tanhf', math.tanh)]
+
+binarys = []
+binarys += [('copysign', 'copysignf', math.copysign)]
+binarys += [('atan2', 'atan2f', math.atan2)]
+binarys += [('pow', 'powf', math.pow)]
+binarys += [('fmod', 'fmodf', math.fmod)]
+binarys += [('hypot', 'hypotf', math.hypot)]
+
+
+@lower(math.isinf, types.Integer)
+@lower(math.isnan, types.Integer)
+def math_isinf_isnan_int(context, builder, sig, args):
+    return context.get_constant(types.boolean, 0)
+
+
+@lower(math.isfinite, types.Integer)
+def math_isfinite_int(context, builder, sig, args):
+    return context.get_constant(types.boolean, 1)
+
+
+def impl_boolean(key, ty, libfunc):
+    def lower_boolean_impl(context, builder, sig, args):
+        libfunc_impl = context.get_function(libfunc,
+                                            typing.signature(types.int32, ty))
+        result = libfunc_impl(builder, args)
+        return context.cast(builder, result, types.int32, types.boolean)
+
+    lower(key, ty)(lower_boolean_impl)
+
+
+def impl_unary(key, ty, libfunc):
+    def lower_unary_impl(context, builder, sig, args):
+        libfunc_impl = context.get_function(libfunc, typing.signature(ty, ty))
+        return libfunc_impl(builder, args)
+
+    lower(key, ty)(lower_unary_impl)
+
+
+def impl_unary_int(key, ty, libfunc):
+    def lower_unary_int_impl(context, builder, sig, args):
+        if sig.args[0] == int64:
+            convert = builder.sitofp
+        else:
+            convert = builder.uitofp
```

Comment:
What if it's accidentally an `int32` if someone tries to reuse this for something? Would it help typing/lowering drift to only accept 64 bit ints?

Reply:
I've now changed it to only accept 64-bit integers.